### PR TITLE
Generate Sprite

### PR DESCRIPTION
**Pull request by oa495**

### CODE REVIEW

File: packages/@icon-magic/config-reader/src/schemas/config-schema.ts [hunk 1]
```diff
@@ -40,7 +40,15 @@ const buildConfigProperties = {
   }
 };
 
-const distributeConfigProperties = {};
+const distributeConfigProperties = {
+  svg: {
+    type: ['object', null],
+    properties: {
+      spriteName: { type: 'string' },
```
Comment (ramitha) [inline packages/@icon-magic/config-reader/src/schemas/config-schema.ts]: This is also optional, yes?

File: packages/@icon-magic/config-reader/src/schemas/config-schema.ts [hunk 2]
```diff
@@ -40,7 +40,15 @@ const buildConfigProperties = {
...
+      inSprite: { type: 'boolean' }
```
Comment (ramitha) [inline packages/@icon-magic/config-reader/src/schemas/config-schema.ts]: This as well.

File: packages/@icon-magic/distribute/src/index.ts [hunk 3]
```diff
@@ -23,28 +22,34 @@ interface ContentImage {
...
-  flag?: string
+  type: ICON_TYPES = 'all',
+  groupByCategory = true
 ): Promise<void> {
-  LOGGER.debug(`entering distribute with ${flag}`);
+  LOGGER.debug(`entering distribute with ${type}`);
   const iconSet = new IconSet(iconConfig, true);
 
+  if (type === 'svg' || type === 'all') {
+    await createSprite(iconSet, outputPath, groupByCategory);
```
Comment (oa495) [inline packages/@icon-magic/distribute/src/index.ts]: Should this be handled separately from the rest or should we incorporate it into the default case: `distributeSvg`? In `icon-magic` v1, there is sorting of the icons before the sprite is created, idk if that's something we want to do?

File: packages/@icon-magic/distribute/src/index.ts [hunk 4]
```diff
@@ -148,13 +153,108 @@ async function distributeByResolution(icon: Icon, outputPath: string) {
...
+}
+
 /**
  * Moves the svg assets of an icon to the outputPath
  * @param icon icon to distribute
  * @param outputPath path to move to
  * @retuns promise after completion
  */
-async function distributeSvg(icon: Icon, outputPath: string): Promise<void[]> {
+async function distributeSvg(icon: Icon, outputPath: string) {
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/index.ts]: Keep a return type

File: packages/@icon-magic/distribute/src/index.ts [hunk 5]
```diff
@@ -148,13 +153,108 @@ async function distributeByResolution(icon: Icon, outputPath: string) {
...
+  for (const icon of iconSet.hash.values()) {
+    LOGGER.debug(`creating sprite with ${icon.iconName}`);
+    if (icon.distribute && icon.distribute.svg && icon.distribute.svg.inSprite) {
+      spriteName = icon.distribute.svg.spriteName || 'icons';
+      const spriteAssets = getIconFlavorsByType(icon, 'svg');
+      for (const asset of spriteAssets) {
+        await appendToSvgDoc(asset, DOCUMENT, svgEl, groupByCategory && icon.category ? icon.category : '');
+      }
+    }
+  }
```
Comment (oa495) [inline packages/@icon-magic/distribute/src/index.ts]: do we assume the spriteName will be the same for now? or even that we want them all in the same svg document?

File: packages/@icon-magic/distribute/test/index-test.ts [hunk 6]
```diff
@@ -1,7 +1,22 @@
...
+
+const FIXTURES = path.resolve(__dirname, '..', '..', 'test', 'fixtures');
+const input = path.resolve(FIXTURES, 'input');
+const output = path.resolve(FIXTURES, 'out');
+const iconSet = configReader.getIconConfigSet(new Array(input));
 
 describe('Test test', function() {
   it('Moves all .png files to the output directory', async () => {
+    await distributeByType(iconSet, output, 'png', false);
+    assert.ok(true, 'Accesses exports');
```
Comment (ramitha) [inline packages/@icon-magic/distribute/test/index-test.ts]: Can we assert the actual set of files and the folder structure that's created here?

File: packages/@icon-magic/distribute/test/index-test.ts [hunk 7]
```diff
@@ -1,7 +1,22 @@
...
+  });
+
+  it('creates the sprite', async () => {
+    await distributeByType(iconSet, output, 'svg', true);
     assert.ok(true, 'Accesses exports');
```
Comment (ramitha) [inline packages/@icon-magic/distribute/test/index-test.ts]: Can we assert that it creates a single file with the correct name and it's contents to contain all the icons specified?

File: packages/@icon-magic/distribute/test/index-test.ts [hunk 8]
```diff
@@ -1,7 +1,22 @@
...
+import * as path from 'path';
+
+import { distributeByType } from '../src';
+
+const FIXTURES = path.resolve(__dirname, '..', '..', 'test', 'fixtures');
+const input = path.resolve(FIXTURES, 'input');
+const output = path.resolve(FIXTURES, 'out');
+const iconSet = configReader.getIconConfigSet(new Array(input));
 
 describe('Test test', function() {
```
Comment (ramitha) [inline packages/@icon-magic/distribute/test/index-test.ts]: Be more descriptive in the test

File: packages/@icon-magic/icon-models/src/interface.ts [hunk 9]
```diff
@@ -17,6 +17,11 @@ export interface AssetConfig {
   contents?: Content;
 }
 
+interface SVGOptions {
+  inSprite: boolean;
```
Comment (ramitha) [inline packages/@icon-magic/icon-models/src/interface.ts]: Do you mean isSprite or inSprite? I prefer a more descriptive variable here.

File: packages/@icon-magic/distribute/src/index.ts [hunk 10]
```diff
@@ -148,13 +153,108 @@ async function distributeByResolution(icon: Icon, outputPath: string) {
...
 
+/**
+ * Creates an SVG Document and sets its attributes
+ * @retuns object with created SVG Document and its child svg element
+ */
+function createSVGDoc(): { DOCUMENT: Document, svgEl: SVGSVGElement } {
+  const SVG_NS = 'http://www.w3.org/2000/svg';
+  const DOCUMENT = new DOMImplementation().createDocument(SVG_NS, 'svg', null);
+  const svgEl = DOCUMENT.createElementNS(SVG_NS, 'svg');
+  svgEl.setAttribute('width', '24px');
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/index.ts]: Why the width and height attributes on the document?

File: packages/@icon-magic/distribute/src/index.ts [hunk 11]
```diff
@@ -148,13 +153,108 @@ async function distributeByResolution(icon: Icon, outputPath: string) {
...
+  svgEl.setAttribute('height', '390px');
+  svgEl.setAttribute('id', 'svg-source');
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/index.ts]: Don't think we need the id either

File: packages/@icon-magic/distribute/src/index.ts [hunk 12]
```diff
@@ -148,13 +153,108 @@ async function distributeByResolution(icon: Icon, outputPath: string) {
...
+  const xml = doc.parseFromString(contents as string, "image/svg+xml");
+  parent.appendChild(xml.documentElement);
+}
+
+/**
+ * Appends icon to element
+ * @param parent element to append to
+ * @param asset the asset whose contents need to be added
+ */
+async function appendToSvgDoc(asset: Asset, doc: Document, svgEl: SVGSVGElement, category: string) {
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/index.ts]: Add return types to all functions

File: packages/@icon-magic/distribute/src/index.ts [hunk 13]
```diff
@@ -148,13 +153,108 @@ async function distributeByResolution(icon: Icon, outputPath: string) {
...
+  if (category) {
+    let def = doc.getElementById(category);
+    if (!def) {
+      def = createDefs(doc, category);
+      svgEl.appendChild(def);
+    }
+    await appendIcon(def, asset);
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/index.ts]: Do we need to await or can we return a promise itself?

File: packages/@icon-magic/distribute/src/index.ts [hunk 14]
```diff
@@ -23,28 +22,34 @@ interface ContentImage {
...
+  LOGGER.debug(`entering distribute with ${type}`);
   const iconSet = new IconSet(iconConfig, true);
 
+  if (type === 'svg' || type === 'all') {
+    await createSprite(iconSet, outputPath, groupByCategory);
+  }
   for (const icon of iconSet.hash.values()) {
-    switch (flag) {
-      case 'createImageSet': {
+    switch (type) {
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/index.ts]: We will have to take care of the case for `all` even for pngs and webps. We could also add a test to check this behavior.

File: packages/@icon-magic/distribute/src/index.ts [hunk 15]
```diff
@@ -148,13 +153,108 @@ async function distributeByResolution(icon: Icon, outputPath: string) {
...
+  LOGGER.debug(`creating svg document ${DOCUMENT}`);
+  return { DOCUMENT, svgEl };
+}
+
+/**
+ * Writes SVG to disk
+ * @param doc svg element to write
+ * @param filePath path to write to
+ */
+function writeSVGToDisk(svgEl: SVGSVGElement, filePath: string){
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/index.ts]: There should be a util in `files` already for writing files.

File: packages/@icon-magic/distribute/src/index.ts [hunk 16]
```diff
@@ -148,13 +153,108 @@ async function distributeByResolution(icon: Icon, outputPath: string) {
...
+ * Creates a sprite and appends SVG icons
+ * @param iconSet set of icons to be added to the sprite
+ * @param outputPath path to write sprite to
+ * @param groupByCategory (for sprite creation) whether to group by the category attribute
+ */
+async function createSprite(iconSet: IconSet, outputPath: string, groupByCategory: boolean) {
+  const { DOCUMENT, svgEl } = createSVGDoc();
+  let spriteName;
+  for (const icon of iconSet.hash.values()) {
+    LOGGER.debug(`creating sprite with ${icon.iconName}`);
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/index.ts]: More verbose in the logs. `adding ${icon.iconName} to ${spriteName}`

File: packages/@icon-magic/distribute/src/index.ts [hunk 17]
```diff
@@ -148,13 +153,108 @@ async function distributeByResolution(icon: Icon, outputPath: string) {
   return Promise.all(promises);
 }
 
+/**
+ * Creates an SVG Document and sets its attributes
+ * @retuns object with created SVG Document and its child svg element
+ */
+function createSVGDoc(): { DOCUMENT: Document, svgEl: SVGSVGElement } {
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/index.ts]: Can we move all the sprite functions into a separate file?

File: packages/@icon-magic/distribute/src/index.ts [hunk 18]
```diff
@@ -170,21 +170,29 @@ async function distributeByResolution(icon: Icon, outputPath: string) {
...
   for (const icon of iconSet.hash.values()) {
     LOGGER.debug(`adding ${icon.iconName} to ${spriteName}`);
-    if (icon.distribute && icon.distribute.svg && icon.distribute.svg.inSprite) {
+    if (icon.distribute && icon.distribute.svg && icon.distribute.svg.toSprite) {
       const iconSpriteName = icon.distribute.svg.spriteName;
       spriteName = iconSpriteName ? iconSpriteName: spriteName;
-      const spriteAssets = getIconFlavorsByType(icon, FILE_TYPE);
-      for (const asset of spriteAssets) {
-        await appendToSvgDoc(asset, DOCUMENT, svgEl, groupByCategory && icon.category ? icon.category : '');
+      if (!spriteNames[spriteName]) {
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/index.ts]: What about the else case here? When it does exist, we've to append to the existing doc.

File: packages/@icon-magic/distribute/src/index.ts [hunk 19]
```diff
@@ -170,21 +170,29 @@ async function distributeByResolution(icon: Icon, outputPath: string) {
...
+        }
+        spriteNames[spriteName] = svgEl;
       }
     }
   }
-  await saveContentToFile(outputPath, spriteName, convertSVGToString(svgEl), FILE_TYPE);
+
+  for (spriteName in spriteNames) {
+    const svgEl = spriteNames[spriteName];
+    await saveContentToFile(outputPath, spriteName, convertSVGToString(svgEl), FILE_TYPE);
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/index.ts]: We can have an array of promises here and not have each one wait until the other completes.

File: packages/@icon-magic/distribute/src/index.ts [hunk 20]
```diff
@@ -218,7 +226,7 @@ async function distributeSvg(icon: Icon, outputPath: string): Promise<void[]>  {
  * @param type The type to which to filter the icon's flavors by
  * @returns a list of flavors that contain assets of "type"
  */
-function getIconFlavorsByType(icon: Icon, type: FlavorType): Asset[] {
+export function getIconFlavorsByType(icon: Icon, type: FlavorType): Asset[] {
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/index.ts]: Remove the export

File: packages/@icon-magic/distribute/test/fixtures/out/filled-1_filled-24x12.imageset/Contents.json [hunk 21]
```diff
@@ -0,0 +1,9 @@
+{
+  "images": [
+    {
+      "idiom": "universal",
+      "scale": "4x",
```
Comment (ramitha) [inline packages/@icon-magic/distribute/test/fixtures/out/filled-1_filled-24x12.imageset/Contents.json]: We should have other scales too. I can debug into this after you submit the PR, could you create an issue if you're seeing only one here?

File: packages/@icon-magic/distribute/test/fixtures/out/icons-1.svg [hunk 22]
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><defs id="ui-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32" id="home-filled-1" data-supported-dps="8x8 16x16" fill="currentColor"><path d="M28 13.36L16.64 6.19a1.2 1.2 0 0 0-1.28 0L4 13.34l1 1.59 2-1.25V25a1 1 0 0 0 1 1h6v-5h4v5h6a1 1 0 0 0 1-1V13.67L27 15z"/></svg></defs><defs id="ui-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32" id="home-filled-1" data-supported-dps="8x8 16x16" fill="currentColor"><path d="M28 13.36L16.64 6.19a1.2 1.2 0 0 0-1.28 0L4 13.34l1 1.59 2-1.25V25a1 1 0 0 0 1 1h6v-5h4v5h6a1 1 0 0 0 1-1V13.67L27 15z"/></svg></defs></svg>
```
Comment (ramitha) [inline packages/@icon-magic/distribute/test/fixtures/out/icons-1.svg]: Can this be prettier looking? New lines, etc.

File: packages/@icon-magic/distribute/src/index.ts (hunk 20 above)
Comment (oa495) [reply packages/@icon-magic/distribute/src/index.ts]: I'm using in the test.

File: packages/@icon-magic/distribute/README.md [hunk 23]
```diff
@@ -0,0 +1,104 @@
+# @icon-magic/distribute
+
+This package is responsible for generating the flavors of the icon in all the
```
Comment (ramitha) [inline packages/@icon-magic/distribute/README.md]: Update the description to what distribute is responsible for. Also need to update the CLI documentation for the distribute command

File: packages/@icon-magic/cli/README.md [hunk 24]
```diff
@@ -52,6 +52,8 @@ Options:
   -c, --config        Path to config file name
   -i, --inputPath     Path to the input directory of icons
   -o, --outputPath    Path to the output directory where the generated assets are to be written to
+  -t, --type          type of icons format to handle, accepted types are svg|png|webp
+  -g, --groupByCategory [for web sprite creation] if to how to group the icons by category
```
Comment (ramitha) [inline packages/@icon-magic/cli/README.md]: nit: Keep indentation uniform

File: packages/@icon-magic/cli/src/index.ts [hunk 25]
```diff
@@ -68,7 +68,7 @@ program
   )
   .option(
     '-g, --groupByCategory',
-    '[for web icons] how to group the icons'
+    '[for web sprite creation] if to how to group the icons by category'
```
Comment (ramitha) [inline packages/@icon-magic/cli/src/index.ts]: if to how to?

File: packages/@icon-magic/distribute/README.md [hunk 26]
```diff
@@ -1,104 +1,86 @@
...
-            "path": "./filled-24x24.png"
-        },
-        "webp": {
-            "name": "filled-24x24",
-            "path": "./filled-24x24.webp"
-        }
-    }
-},
-```
+This package is responsible for massaging the assets into type specific structures and formats. At the end of the `generate` step, we have optimized .svg files, .png files and .webp files for web, ios and android consumption. You can look in `test/fixtures/input` for an example of the file structure after the generate step. `distribute` organizes and structures those files how they need to be consumed and creates the necessary files for platform consumption.
```
Comment (ramitha) [inline packages/@icon-magic/distribute/README.md]: You can add a link to the file in the readme. I believe the syntax is `[I'm a relative reference to a repository file](../blob/master/LICENSE)`

File: packages/@icon-magic/distribute/src/create-sprite.ts [hunk 27]
```diff
@@ -55,7 +74,10 @@ async function appendIcon(parent: Element, asset: Asset): Promise<void> {
  */
 export async function appendToSvgDoc(asset: Asset, doc: Document, svgEl: SVGSVGElement, category: string): Promise<void> {
   if (category) {
-    let def = doc.getElementById(category);
+    // TODO: Replace this with findElementById, which right now doesn't find the <defs> with the ID
```
Comment (ramitha) [inline packages/@icon-magic/distribute/src/create-sprite.ts]: Is it not supported in xmldom?

Comment (oa495) [reply packages/@icon-magic/distribute/src/create-sprite.ts]: It is, and the method calls, it just returns `null` every time, not sure why.

File: packages/@icon-magic/distribute/src/create-sprite.ts [hunk 28]
```diff
@@ -0,0 +1,94 @@
...
+  svgEl.setAttribute('version', '1.1');
+  DOCUMENT.appendChild(svgEl);
+  LOGGER.debug(`creating svg document ${DOCUMENT}`);
+  return { DOCUMENT, svgEl };
+}
+
+/**
+ * Looks for <defs> elements with an ID in Document
+ * @param doc the Document to look in
+ * @param the category to look for as an ID attribute
```
Comment (zimmi88) [inline packages/@icon-magic/distribute/src/create-sprite.ts]: `@param category ...`

File: packages/@icon-magic/distribute/src/index.ts [hunk 29]
```diff
@@ -23,31 +28,45 @@ interface ContentImage {
 interface AssetCatalog {
   images?: ContentImage[];
 }
+type ICON_TYPES = 'svg' | 'png' | 'webp' | 'all';
```
Comment (zimmi88) [inline packages/@icon-magic/distribute/src/index.ts]: Should this be an enum?

**Comment by oa495** [comment]
Need to open

**Pull request merged**